Update AutoSuspend v1.2.1

jurassicplayer opened the pull request:
# AutoSuspend v1.2.1
- Reorganized all code
   - Re-implemented a slightly modified copy of decky-loader toast to allow toggle-able sound (temp fix until decky-loader can handle that internally)
       - Allows for Toast Only / Sound Only / Toast & Sound / No Toast or Sound
       - Should still follow the SteamOS "Disable non-critical notifications" and "Hide non-critical notifications while in-game" settings
   - Re-arranged layout to have the actual plugin functions at the top
   - Now uses an autosuspend.json configuration in the homebrew/settings folder ***

*** NOTE: decky-loader currently has a small bug that causes the homebrew/settings folder to have unintended folder permissions.
Please only use this release after either:
A. Upstream decky-loader has fixed the issue (https://github.com/SteamDeckHomebrew/decky-loader/issues/257) or,
B. Using a terminal or ssh connection, run `sudo chown deck:deck ~/homebrew/settings`

## Checklist:

- [x] I am the original author of this plugin.
- [x] I made my code efficient and readable.
- [x] I have verified that my plugin works properly on the latest versions of SteamOS and decky-loader.
- [x] I have verified that my plugin is unique and does not have the same functionality as another plugin on the store.

TrainDoctor commented:
As already noted, once we merge [#257](https://github.com/SteamDeckHomebrew/decky-loader/issues/257) then I'll have testers begin their testing for this plugin.

TrainDoctor commented:
> As already noted, once we merge [#257](https://github.com/SteamDeckHomebrew/decky-loader/issues/257) then I'll have testers begin their testing for this plugin.

Approved for testing on latest stable/pre-releases of decky loader on SteamOS branches; Stable, Beta and Preview.

BottomNotch commented:
So, here's what I've found with SteamOS main, Steam Client Beta, Decky pre-release.

The most serious issue, which I'm unfortunately not sure how to reproduce, the critical level setting seems to have gotten stuck at 10%, I've changed it to 5% and it appears to have changed, but when it drops bellow 10% (which is now what I've set my warning to) I get the warning notification immediately followed by the suspend notification.  I wasn't initially having any trouble changing this setting when I began testing.

Here's the minor things I've encountered:
1. The warning and suspend only happen when you get bellow the set threshold, not when you reach the threshold, perhaps this is intentional, but...
2. If you are bellow either threshold, and to charge to exactly one of the thresholds, but not past it and you then discharge below the threshold, nothing happens.
3. The notification settings may not be working, but I'm not sure because I starting testing the at about the same time I was having issues changing the critical level
4. You don't have to "apply changes" for the notification settings to apparently change, I don't know if they actually change without applying changes, or if the switch position just persists, It's impossible to tell because I can't seem to get the notification behavior to change at all, wether or not I apply changes.

jurassicplayer commented:
> The most serious issue, which I'm unfortunately not sure how to reproduce, the critical level setting seems to have gotten stuck at 10%, I've changed it to 5% and it appears to have changed, but when it drops bellow 10% (which is now what I've set my warning to) I get the warning notification immediately followed by the suspend notification.
>     1. The warning and suspend only happen when you get bellow the set threshold, not when you reach the threshold, perhaps this is intentional, but...
> 
>     2. If you are bellow either threshold, and to charge to exactly one of the thresholds, but not past it and you then discharge below the threshold, nothing happens.
> 
>     3. The notification settings may not be working, but I'm not sure because I starting testing the at about the same time I was having issues changing the critical level
> 
>     4. You don't have to "apply changes" for the notification settings to apparently change, I don't know if they actually change without applying changes, or if the switch position just persists, It's impossible to tell because I can't seem to get the notification behavior to change at all, wether or not I apply changes.

For minor things 1. and 2., that is the intended behavior in order to prevent accidentally suspending/notifying more than once in any given scenario where the threshold has been passed. Kind of like a "beyond a reasonable doubt" that the notification/suspend should fire again.

For 4., the notification settings are indeed applied instantly (one of the reasons why I moved it below the `Apply Changes` button was to signify that it wasn't related). Threshold levels aren't applied instantly for the simple reason of "I don't want my console to suspend while I'm fiddling with settings" and I have no way of detecting when a user has completed their change.

I believe the serious issue and 3/4 are related (all of the settings are handled the same way, so if one is screwed up, all of them should be), but from what I know about javascript (which is admittedly not much), it shouldn't be possible for the warning indicator to trigger first before the critical in a scenario where the battery level is below both thresholds because the first check I make is if the battery level is past critical. After noticing the issue, have you done a full system restart and does it resolve the issue or is it still present? I've been trying to reproduce the issue, but I haven't been able to yet.

This did indirectly lead me to a separate minor bug: Going from both warning/critical below battery level to both above with warning level still above critical causes the warning notification to appear after the critical notification (going to fix this). Ex. warning 20/critical 10/battery 60% -> warning 80/critical 75/battery 60%

jurassicplayer commented:
@BottomNotch and any other testers that run into issues, if you could re-test with this version, hopefully reproduce the issue, and attach the CEF debugging logs, that would be great.

AAGaming00 commented:
oops, misclick

GeneStarwind09 reviewed (commented):
Tested on beta, working without any noticeable issues

TrainDoctor commented:
Requires preview testing.

OuterWinnie commented:
Haven't found any problem on preview 3.4
